reporter/collector_reporter.go

```diff
@@ -0,0 +1,491 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package reporter // import "go.opentelemetry.io/ebpf-profiler/reporter"
+
+import (
+	"context"
+	"maps"
+	"slices"
+	"time"
+
+	lru "github.com/elastic/go-freelru"
+	log "github.com/sirupsen/logrus"
+	"go.opentelemetry.io/collector/consumer/consumerprofiles"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pprofile"
+	semconv "go.opentelemetry.io/otel/semconv/v1.25.0"
+
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/libpf/xsync"
+)
+
+// Assert that we implement the full Reporter interface.
+var _ Reporter = (*CollectorReporter)(nil)
+
+// OTLPReporter receives and transforms information to be Collector Collector compliant.
+type CollectorReporter struct {
+	cfg          *Config
+	nextConsumer consumerprofiles.Profiles
+
+	// stopSignal is the stop signal for shutting down all background tasks.
+	stopSignal chan struct{}
+
+	// hostmetadata stores metadata that is sent out with every request.
+	hostmetadata *lru.SyncedLRU[string, string]
+
+	// executables stores metadata for executables.
+	executables *lru.SyncedLRU[libpf.FileID, execInfo]
+
+	// frames maps frame information to its source location.
+	frames *lru.SyncedLRU[libpf.FileID, *xsync.RWMutex[map[libpf.AddressOrLineno]sourceInfo]]
+
+	// traceEvents stores reported trace events (trace metadata with frames and counts)
+	traceEvents xsync.RWMutex[map[traceAndMetaKey]*traceEvents]
+
+	// samplesPerSecond is the number of samples per second.
+	samplesPerSecond int
+}
+
+// NewCollector builds a new CollectorReporter
+func NewCollector(cfg *Config, nextConsumer consumerprofiles.Profiles) (*CollectorReporter, error) {
+	executables, err :=
+		lru.NewSynced[libpf.FileID, execInfo](cfg.ExecutablesCacheElements, libpf.FileID.Hash32)
+	if err != nil {
+		return nil, err
+	}
+	executables.SetLifetime(1 * time.Hour) // Allow GC to clean stale items.
+
+	frames, err := lru.NewSynced[libpf.FileID,
+		*xsync.RWMutex[map[libpf.AddressOrLineno]sourceInfo]](
+		cfg.FramesCacheElements, libpf.FileID.Hash32)
+	if err != nil {
+		return nil, err
+	}
+	frames.SetLifetime(1 * time.Hour) // Allow GC to clean stale items.
+
+	// Next step: Dynamically configure the size of this LRU.
+	// Currently, we use the length of the JSON array in
+	// hostmetadata/hostmetadata.json.
+	hostmetadata, err := lru.NewSynced[string, string](115, hashString)
```

Comment:
Not sure what the rationale was for choosing an LRU here instead of a map 🤔 
We don't have to change it in this PR, though.

Reply:
It's just the same than we use in the OTLP reporter.

Reply:
I know, same thoughts on that code. Here, we just copied technical debt. But I am just raising this for awareness, it's not meant as a blocker.